### ISSUE TITLE
OBC-PEER Crashes with "invalid memory address or nil pointer dereference" in ocbpbft.moveWatermarks

### ISSUE THREAD

**Issue by bcbrock** [hyperledger-archives/fabric#696]
@jyellick   @kchristidis 

Occurs in code with #687 fix (plus others) 

```
commit 7abd041f5b9e5ce346d1608f26b6880f6e235a77
Merge: 5077256 4fd8ae5
Author: Binh Q Nguyen <binhn@us.ibm.com>
Date:   Wed Feb 17 13:29:30 2016 -0500

    Merge pull request #682 from adecaro/crash_issue669

    Fix for issue #669
```

Same setup and irritation as in #687. I have not modified PBFT parameters. VP3 crashed in this run:

```
[36m21:47:27.441 [consensus/obcpbft] recvRequest -> DEBU 15115[0m Replica 3 received request: xLiYonKjyP43bjueApdggwgGibL6F3R7UcZbEaVt23vNZZX5qlFBcq3pr54J3Lowz1HMxxyY3MP9MGk9jBqWQA==
[36m21:47:27.446 [consensus/obcpbft] recvRequest -> DEBU 15116[0m Replica 3 received request: 1LokxEyMUyRCAtTX8D5bEtO4mF1ctU4o0Vqn11rTO+oEdOSTSiiX1NcJ/scPGpJRZ1ggWzcI3WZHzmxOkgkQGw==
[36m21:47:27.449 [consensus/obcpbft] recvRequest -> DEBU 15117[0m Replica 3 received request: oSrrJbXcOfiqPdXlPsvSnapoAd6M8UNeql9dyf9wIk/NnSUnecMu3hwcSL0ss2Lcm6os5NCJasQ+0/VynmHmOw==
[36m21:47:27.450 [consensus/obcpbft] recvCheckpoint -> DEBU 15118[0m Replica 3 received checkpoint from replica 1, seqNo 470, digest rnI0/gpXpbBA8S4EcL1IN6yozzzANXehPXS9kfFFUZyXphbLLKljR5aeCRE91Fd1eH6hGOQ9BDqsnDmDayooXg==
[33m21:47:27.450 [consensus/obcpbft] recvCheckpoint -> WARN 15119[0m Checkpoint sequence number outside watermarks: seqNo 470, low-mark 440
[36m21:47:27.452 [consensus/obcpbft] recvCheckpoint -> DEBU 1511a[0m Replica 3 received checkpoint from replica 2, seqNo 470, digest rnI0/gpXpbBA8S4EcL1IN6yozzzANXehPXS9kfFFUZyXphbLLKljR5aeCRE91Fd1eH6hGOQ9BDqsnDmDayooXg==
[33m21:47:27.452 [consensus/obcpbft] witnessCheckpoint -> WARN 1511b[0m Replica 3 is out of date, f+1 nodes agree checkpoint with seqNo 470 exists but our high water mark is 460
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 1511c[0m Replica 3 cleaning quorum certificate for view=0/seqNo=446
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 1511d[0m Replica 3 cleaning quorum certificate for view=0/seqNo=448
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 1511e[0m Replica 3 cleaning quorum certificate for view=0/seqNo=445
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 1511f[0m Replica 3 cleaning quorum certificate for view=0/seqNo=449
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15120[0m Replica 3 cleaning quorum certificate for view=0/seqNo=456
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15121[0m Replica 3 cleaning quorum certificate for view=0/seqNo=457
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15122[0m Replica 3 cleaning quorum certificate for view=0/seqNo=459
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15123[0m Replica 3 cleaning quorum certificate for view=0/seqNo=447
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15124[0m Replica 3 cleaning quorum certificate for view=0/seqNo=450
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15125[0m Replica 3 cleaning quorum certificate for view=0/seqNo=455
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15126[0m Replica 3 cleaning quorum certificate for view=0/seqNo=458
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15127[0m Replica 3 cleaning quorum certificate for view=0/seqNo=460
[36m21:47:27.452 [consensus/obcpbft] moveWatermarks -> DEBU 15128[0m Replica 3 cleaning quorum certificate for view=0/seqNo=444
[36m21:47:27.452 [peer] DeregisterHandler -> DEBU 15129[0m Deregistered handler with key: name:"vp2" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0xa4f126]

goroutine 40 [running]:
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*pbftCore).moveWatermarks(0xc82012d900, 0x1d6)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/pbft-core.go:879 +0xb36
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*pbftCore).witnessCheckpoint(0xc82012d900, 0xc824d0b350)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/pbft-core.go:948 +0x50e
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*pbftCore).recvCheckpoint(0xc82012d900, 0xc824d0b350, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/pbft-core.go:1003 +0x271
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*pbftCore).recvMsgSync(0xc82012d900, 0xc8202eb4c0, 0x2, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/pbft-core.go:528 +0xe24
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*pbftCore).receive(0xc82012d900, 0xc8252b59d0, 0x64, 0x64, 0x2, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/pbft-core.go:487 +0x23e
github.com/openblockchain/obc-peer/openchain/consensus/obcpbft.(*obcClassic).RecvMsg(0xc820317dc0, 0xc824fbcf80, 0xc8203fbd00, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/obcpbft/obc-classic.go:70 +0x548
github.com/openblockchain/obc-peer/openchain/consensus/helper.(*ConsensusHandler).HandleMessage(0xc8202e0b70, 0xc824fbcf80, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/consensus/helper/handler.go:74 +0x109
github.com/openblockchain/obc-peer/openchain/peer.(*PeerImpl).handleChat(0xc8201f2540, 0x7f35c3194948, 0xc820016650, 0x7f35c3195028, 0xc8203d0830, 0x1, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/peer/peer.go:636 +0x679
github.com/openblockchain/obc-peer/openchain/peer.(*PeerImpl).chatWithPeer(0xc8201f2540, 0xc820352cd0, 0x10, 0x0, 0x0)
    /go/src/github.com/openblockchain/obc-peer/openchain/peer/peer.go:611 +0x9f5
created by github.com/openblockchain/obc-peer/openchain/peer.(*PeerImpl).PeersDiscovered
    /go/src/github.com/openblockchain/obc-peer/openchain/peer/peer.go:321 +0x36e
```

Full log files. Note the other peers kept running after VP3 crashed.
[logs.zip](https://github.com/openblockchain/obc-peer/files/135258/logs.zip)

**Comment by bcbrock** [comment]
~~This failure seems to be rare. Since this single occurrence the same tests have all worked (modulo #699).~~ Increasing the number of transactions hits it every time.

**Comment by jyellick** [comment]
@kchristidis Here is a working theory for how this might be happening:

For whatever reason, we receive a commit message for a request which we have not received a preprepare.  Then the code enters into `recvCommit`.  The call:

```
                cert := instance.getCert(commit.View, commit.SequenceNumber)
```

is made, which internally, is this:

```
        idx := msgID{v, n}
        cert, ok := instance.certStore[idx]
        if ok {
                return
        }

        cert = &msgCert{}
        instance.certStore[idx] = cert
        return
```

So, because there is no `msgCert` in `certStore`, we create and insert one.  Now, there is a `msgCert` in `certStore` which has a `nil` `prePrepare` field.  Consequently, when we try to dereference `cert.prePrepare.RequestDigest` as seen in the logs above, we get a `nil` pointer dereference error.

One solution would be to check both the `prePrepare`, `prepare`, and `commit` fields for non-`nil`, and extract the request digest from whichever is present.  This would fix the bug, but the behavior seems inconsistent to me.  Edit: This seems like a bad plan.  An additional concern here would be that if we delete requests for which we have commit messages, a malicious peer could essentially send commit messages as it receives requests, causing those requests to be deleted at the next stable checkpoint, which could cause all sorts of problems.

~~~We discard prepares for requests which are not prePrepared, why are we storing commits which are not prePrepared or prepared? (I can see the advantage to storing them, in case things are received out of order, but then it seems that too, we should store the prepares).  Can you explain what's going on here @kchristidis~~~

My mistake above, the prepare behaves like the commit.

**Comment by kchristidis** [comment]
Reviewing the code for this.

**Comment by kchristidis** [comment]
(Clarified this with Jason on Slack, but posting here for reference as well. We do _not_ discard prepares for requests which are not prePrepared.)

**Comment by kchristidis** [comment]
There's an `if pre-prepare not found, discard prepare (or commit)` logic we need to integrate.

I've added it to the `recvPrepare` and `recvCommit` functions already (https://github.com/kchristidis/obc-peer/commits/issues/696), but following the conversation Jason and I had with Marko and Jason's comments, I think this now probably belongs to the `moveWatermarks` routine.

Jason if I'm wrong, let me know.

**Comment by jyellick** [comment]
@kchristidis I thought from our conversation with @vukolic that it is fine to to accept commits/prepares for requests which we did not receive a preprepare, in fact, this is a valid optimization.

Instead, moving the watermarks in the event of state transfer should have the special property of discarding all requests, and waiting for state transfer to catch us up.  Because we will be receiving prePrepares for all requests after the sequence number which state transfer recovers to, this discarding is safe.

Ultimately, to ensure that requests are not discarded in a segregated network, #394 will need to be implemented, but that is outside the scope of this bug.

**Comment by bcbrock** [comment]
I will be out of the office until Monday. If you want to test a fix the test procedure is given in #687.

**Comment by bcbrock** [comment]
This one seems to be fixed.
